pkg/datapath/linux/bandwidth/ops.go

```diff
@@ -57,19 +57,30 @@ func (ops *ops) Update(ctx context.Context, txn statedb.ReadTxn, q *tables.Bandw
 	if err != nil {
 		return fmt.Errorf("QdiscList: %w", err)
 	}
-	if len(qdiscs) > 0 {
-		ok := qdiscs[0].Type() == "mq"
-		if qdiscs[0].Type() == "fq" {
-			fq, _ := qdiscs[0].(*netlink.Fq)
+	mqSeen := false
+	qdiscUpdateNeeded := false
+	for _, qdisc := range qdiscs {
+		if qdisc.Type() == "mq" && qdisc.Attrs().Parent == netlink.HANDLE_ROOT {
+			if mqSeen {
+				qdiscUpdateNeeded = true
+				break
+			}
+			mqSeen = true
+		}
+		if qdisc.Type() == "fq" {
+			fq, _ := qdisc.(*netlink.Fq)
 
 			// If it's "fq" and with our parameters, then assume this was
 			// already set up and there wasn't any MQ support.
-			ok = fq.Horizon == uint32(q.FqHorizon.Microseconds()) &&
+			ok := fq.Horizon == uint32(q.FqHorizon.Microseconds()) &&
 				fq.Buckets == q.FqBuckets
+			if !ok {
+				qdiscUpdateNeeded = true
+			}
```

Comment:
this still has an issue, meaning, if there is neither mq nor fq, then qdiscUpdateNeeded remains false. I think it's better to reverse the logic and starting out with qdiscUpdateNeeded true, then iterating qdiscs. We may find mq with fq leaf qdiscs or just single fq qdisc as root one. If not, then we need to update. If yes, then we need to look at fq params. If they don't match, we need to update, otherwise no work needed.

Reply:
Also Cc'ing @joamaki fyi.

Reply:
Makes sense. Will update this.